src/v/model/ktp.h

```diff
@@ -0,0 +1,347 @@
+
+    /**
+     * @brief Compare this object with an ntp.
+     *
+     * Although this must compare the namespaces it is still in practice faster
+     * than an ntp <-> ntp comparison, because one side of the namespace
+     * comparison the constexpr string "kafka", which leads to very efficient
+     * code generation (an inlined, fixed-size memcmp) for that comparison.
+     */
+    bool operator==(const ntp& other) const {
+        return tp_equals(other.tp, tp) && kafka_ns_view == other.ns();
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const ktp& t) {
+        os << t.to_ntp();
+        return os;
+    }
+
+private:
+    /**
+     * @brief Helper to compare two topic_partition objects efficiently.
+     *
+     * When using libc++, comparing two ss::sstring objects uses a relatively
+     * slow byte-by-byte comparison, because ss::string uses std::equal_to with
+     * a character range, and this is not recognized as a memcmp (though similar
+     * patternrs are). By explicitly using std::string_view::operator==, as
+     * here, we get a recognized memcmp and a generally faster comparison.
+     */
+    static bool tp_equals(
+      const model::topic_partition& l, const model::topic_partition& r) {
+        return l.partition == r.partition
+               && std::equal_to<std::string_view>{}(l.topic(), r.topic());
+    }
+};
+} // namespace model
+
+namespace std {
+template<>
+struct hash<model::ktp> {
+    /**
+     * @brief The hash specialization for ktp.
+     *
+     * Compatible with the hash for ntp: a ktp will return the same hash as an
+     * ntp with the same topic, partition and the "kafka" namespace.
+     */
+    size_t operator()(const model::ktp& ktp) const {
+        return detail::ntp_hash(
+          model::kafka_ns_view, ktp.get_topic()(), ktp.get_partition());
+    }
+};
+} // namespace std
+
+namespace model {
+
+/**
+ * @brief A ktp object that remembers its hash code.
+ *
+ * In some contexts, a ktp/ntp object will be repeatedly hashed, which
+ * is a relatively expensive operation. Instead, we can swap in this type
+ * which remembers its hash code. It is a subclass of ktp, and can safely
+ * be passed as a reference to a ktp or sliced down to a ktp for convenience.
+ *
+ * The hash code is calculated once at constructon and never changes, except via
+ * assignment, since the underlying topic and partition cannot be changed (but
+ * this invariant must be enforced by the ktp class as noted there). Among other
+ * things this means that assignment of a ktp_with_hash from a ktp object does
+ * not compile: for this to be safe we'd need to recalculate the hash after
+ * assignment.
+ */
+struct ktp_with_hash : public ktp {
+    /**
+     * @brief Default-construct a ktp_with_hash object.
+     *
+     * Equivalent to an object with empty topic name
+     * and partition id 0.
+     */
+    ktp_with_hash()
+      : ktp_with_hash{"", 0} {}
```

Comment:
It is not possible with boost::hash stuff or abseil hash stuff because neither of them are compile-time functions, and at least with abseil it is explicitly so: they use an address of a global object as the seed for the hash with the idea that ASLR will change the hash for every process even if the binary is the same.

That said, this is important for the case of hashing the string "kafka" as we do for `ktp` and I did check that it is calculated at compile time despite not being constexpr, I expect this constructor with even simpler input "" and 0 should also be calculated at compile time. 

I don't think it's too important for the default constructor as I think this is not often used.

Reply:
@dlex resolving this assuming you are OK with this explanation.